### PR TITLE
WIP - Add automation tests to go-spacemesh

### PR DESCRIPTION
**Pull request by beckmani**
This pull request is for adding the automation tests to go-spacemesh repository.
The reason the tests are in this repository is to enable us create matching between the tests and the code on which they run.
This is still WIP so many changes will be applied to the tests after we add it to go-spacemesh repository

### CODE REVIEW

File: tests/fixtures.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import pytest
+
+
+@pytest.fixture
+def set_namespace(request):
+    ns = getattr(request.module, "NAMESPACE", "aaa")
```
Comment (y0sher) [inline tests/fixtures.py]: what `aaa` ? please explain or parametrize

File: tests/config.yaml [hunk 2]
```diff
@@ -0,0 +1 @@
+namespace: isaac
```
Comment (y0sher) [inline tests/config.yaml]: this will be `isaac` for all the devs all the time ? if not parametrize

File: tests/k8s/oracle.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        name: oracle
+    spec:
+      containers:
+      - name: oracle
+        image: beckmani/oracle_server:latest
```
Comment (antonlerner) [inline tests/k8s/oracle.yml]: please move docker container to a dedicated spacemesh docker repo

Comment (y0sher) [inline tests/k8s/oracle.yml]: parametrize image

File: tests/k8s/client.yml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+       run: client
+  replicas: 5
+  template:
+    metadata:
+      labels:
+        run : client
+    spec:
+      containers:
+      - name: client
+        image: spacemeshos/go-spacemesh:ms1
```
Comment (y0sher) [inline tests/k8s/client.yml]: the image should be parametrized, we can't expect each dev to replace yaml file params each tests he runs. same goes for the `args` below.

File: tests/test_bs.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    resp = k8s_beta.delete_namespaced_deployment(name=deployment_name,
+                                                 namespace=name_space,
+                                                 body=client.V1DeleteOptions(
+                                                     propagation_policy='Foreground',
+                                                     grace_period_seconds=5))
+    return resp
+
+
+def query_bootstrap_es(indx, namespace, bootstrap_po_name):
+    es = Elasticsearch(['http://127.0.0.1:9200'])
```
Comment (y0sher) [inline tests/test_bs.py]: you write that line for each method, looks a little bit bad. 
either use a const for the address, or create `es` somewhere else

File: tests/test_bs.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    k8s_beta = client.ExtensionsV1beta1Api()
+    resp = k8s_beta.delete_namespaced_deployment(name=deployment_name,
+                                                 namespace=name_space,
+                                                 body=client.V1DeleteOptions(
+                                                     propagation_policy='Foreground',
+                                                     grace_period_seconds=5))
+    return resp
+
+
+def query_bootstrap_es(indx, namespace, bootstrap_po_name):
```
Comment (y0sher) [inline tests/test_bs.py:93]: this name seems too similar to the method below but it's really different. consider better naming

File: tests/test_bs.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    hits = list(s.scan())
+
+    print("Number of Hits: {0}".format(len(hits)))
+    s=set([h.N for h in hits])
+    return (len(s))
+
+
+def query_es_gossip_message(indx, namespace, client_po_name):
+    es = Elasticsearch(['http://127.0.0.1:9200'])
+    fltr = Q("match_phrase", kubernetes__namespace_name=namespace) & \
```
Comment (y0sher) [inline tests/test_bs.py:121]: maybe we can take out the pod and namespace comparing to a different func and duplicate a little bit less

File: tests/test_bs.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    hits = list(s.scan())
+
+    print("Number of gossip Hits: {0}".format(len(hits)))
+    s=set([h.N for h in hits])
+    return (len(s))
+
+
+@pytest.fixture(scope='session')
+def load_config():
+    config.load_kube_config('/Users/isaac/.kube/config')
```
Comment (y0sher) [inline tests/test_bs.py]: this path won't work for everyone, maybe add it as param or require the file to be in the execution folder

File: tests/test_bs.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+    request.addfinalizer(fin)
+    return _setup_clients_in_namespace(testconfig['namespace'])
+
+
+#==============================================================================
+#    TESTS
+#==============================================================================
+
+current_index = 'kubernetes_cluster-2019.02.2*'
```
Comment (y0sher) [inline tests/test_bs.py]: If I get the purpose of this right, it's better to just generate the date part by the actual running date

File: tests/test_bs.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        global bs_info
+        bs_info = NodeInfo()
+        resp=create_deployment(BOOT_DEPLOYMENT_FILE, name_space)
+
+        bs_info.bs_deployment_name = resp.metadata._name
+        namespaced_pods = client.CoreV1Api().list_namespaced_pod(namespace=name_space).items
+        bootstrap_pod = next(filter(lambda i: i.metadata.name.startswith(bs_info.bs_deployment_name), namespaced_pods))
+        bs_info.bs_pod_name = bootstrap_pod.metadata.name
+
+        while True:
```
Comment (y0sher) [inline tests/test_bs.py:162]: should probably be taken to a different function as well, (used again below)

File: tests/test_bs.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+
+#==============================================================================
+#    TESTS
+#==============================================================================
+
+current_index = 'kubernetes_cluster-2019.02.2*'
+
+
+def test_bootstrap(load_config, setup_bootstrap):
+    time.sleep(10)
```
Comment (y0sher) [inline tests/test_bs.py]: we do this only to wait for the record to show up in ES right? worth a comment.
also this might be useful in the future - https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-refresh.html

File: tests/kubectl-cmd.sh [hunk 12]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+
+set -xe
+
+export end_point='http://'$1':9090/'$3
+echo $2
+#kubectl run --quiet curl-isaac --image=tutum/curl -i --tty --restart=Never --rm -n sm -- curl --request POST --data '{ "data": "foo" }' $end_point
+kubectl run --quiet curl-isaac --image=tutum/curl -i --tty --restart=Never --rm -n sm -- curl --request POST --data $2 $end_point
```
Comment (y0sher) [inline tests/kubectl-cmd.sh:8]: isn't this file used for running curl commands ? couldn't get it from its name

File: tests/test_bs.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+def test_gossip(load_config, setup_clients):
+    global client_info
+
+    # send message to client via rpc
+    client_ip = setup_clients[0].status.pod_ip
+    print("Sending gossip from client ip: {0}".format(client_ip))
+
+    api = 'v1/broadcast'
+    data = '{"data":"foo"}'
+    p = subprocess.Popen(['./kubectl-cmd.sh', '%s' % client_ip, "%s" % data, api], stdin=subprocess.PIPE, stdout=subprocess.PIPE,
```
Comment (y0sher) [inline tests/test_bs.py]: you should probably wrap the RPC call logic in a function.

**Comment by y0sher** [comment]
just figured out this PR is still WIP, hope my comments help anyway :)

File: tests/k8s/client.yml (hunk 4 above)
Comment (beckmani) [reply tests/k8s/client.yml]: This file is used as a template for the deployment. This parameters have the most general values.
Some of them are override in the test execution with values from a config file (config.yaml)

File: tests/test_bs.py (hunk 13 above)
Comment (beckmani) [reply tests/test_bs.py]: ok

File: tests/k8s/oracle.yml (hunk 3 above)
Comment (beckmani) [reply tests/k8s/oracle.yml]: Done. See:
https://cloud.docker.com/u/spacemeshos/repository/docker/spacemeshos/oracle

File: tests/k8s/bootstrap-w-conf.yml [hunk 14]
```diff
@@ -0,0 +1,26 @@
+apiVersion: extensions/v1beta1
```
Comment (y0sher) [inline tests/k8s/bootstrap-w-conf.yml:1]: that's a little bit of a mess. I see you added this file and the config. but you didn't add `w-conf` for clients. also you kept the old file as well. 
I prefer we won't push half-baked features to develop and we split it to a different pull request.

**Review by y0sher — APPROVED** [review]
looks good generally. a lot of work ahead :)
good job 👍 

don't forget to rebase on develop
